documentation/docs/extensibility.md

```diff
@@ -1,86 +1,184 @@
 
-  ``` groovy
-  void call(Map params) {
-      //access stage name
-      echo "Start - Extension for stage: ${params.stageName}"
+Options for extensibility, in order in which we recommend considering them:
 
-      //access config
-      echo "Current stage config: ${params.config}"
+## 1) Extend individual stages
 
-      //execute original stage as defined in the template
-      params.originalStage()
+In this option, you use the centrally maintained pipeline, but can change individual stages if required.
 
-      //access overall pipeline script object
-      echo "Branch: ${params.script.commonPipelineEnvironment.gitBranch}"
+To do so, create a file called `<StageName>.groovy` (for example, `Acceptance.groovy` or `lint.groovy`) in `.pipeline/extensions/` in your source code repository.
+
+For this, you need to know the technical identifiers for stage names.
+
+* For the general purpose pipeline, you can find them in [the pipeline source file](https://github.com/SAP/jenkins-library/blob/master/vars/piperPipeline.groovy).
+* For SAP Cloud SDK Pipeline, you can find them in [this GitHub search query](https://github.com/SAP/cloud-s4-sdk-pipeline-lib/search?q=%22def+stageName+%3D%22).
+
+The centrally maintained pipeline checks if such a file exists and executes it, if present.
+A parameter of type `map` that contains the following keys is passed to the extension:
+
+* `script`: defines the global script environment of the `Jenkinsfile` run. This makes sure that the correct configuration environment can be passed to project "Piper" steps and allows access to for example the `commonPipelineEnvironment`.
+* `originalStage`: this will allow you to execute the "original" stage at any place in your script. If omitting a call to `originalStage()` only your code will be executed instead.
+* `stageName`: name of the current stage
+* `config`: configuration of the stage and general config (including all defaults)
+
+Here a simple example for such an extension, which you can use as a starting point:
+
+```groovy
+void call(Map params) {
+  //access stage name
+  echo "Start - Extension for stage: ${params.stageName}"
+
+  //access config
+  echo "Current stage config: ${params.config}"
+
+  //execute original stage as defined in the template
+  params.originalStage()
+
+  //access overall pipeline script object
+  echo "Branch: ${params.script.commonPipelineEnvironment.gitBranch}"
+
+  echo "End - Extension for stage: ${params.stageName}"
+}
+return this
+```
+
+!!! note "`return this`"
+    Don't forget the `return this` which is required at the end of _all_ extension scripts.
+    This is due to how Groovy loads scripts internally.
 
-      echo "End - Extension for stage: ${params.stageName}"
-  }
-  return this
-  ```
 
 !!! note "Init stage cannot be extended"
     Please note, the `Init` stage among other things also checks out your current repository.<br />Thus it is not possible to use extensions on this stage.
```

Comment:
Please note, the `Init` stage among other things also checks out your current repository *including your extensions*.<br />Thus it is not possible to use extensions on this stage.